File: go/heap/295_median_finder.go
package heap

import "container/heap"

// 数据流的中位数
// 中位数将数组分为两个部分，左边的元素比其小，右边的比其大
// 左边大顶堆，右边小顶堆，小的加左边，大的加右边，
// 平衡俩堆数，新加就弹出，堆顶给对家，奇数取多的，偶数取除2.
type MedianFinder struct {
	qMin bigHeap // 左边的元素，大根堆
	qMax sHeap   // 右边的元素，小根堆
}

func Constructor2() MedianFinder {
	qMin := make(bigHeap, 0)
	qMax := make(sHeap, 0)
	heap.Init(&qMin)
	heap.Init(&qMax)

	return MedianFinder{
		qMin: qMin,
		qMax: qMax,
	}
}

// AddNum 添加一个元素
// 如果总元素个数是奇数，qMin 中的数的数量比 qMax 多一个，此时中位数为 queMin 队头
// 当累计添加的数的数量为偶数时，两个优先队列中的数的数量相同，此时中位数为它们的队头的平均值。
func (m *MedianFinder) AddNum(num int) {
	// 当不存在元素的时候 num 加入到qMin 中
	// 或者挡 num < qMin 中的最大值时，加入左边的堆中，并且将此时元素数量发生变化
	// 将qMin 中的堆顶元素(最大值) 移到右边列表
	if m.qMin.Len() == 0 || num < m.qMin[0] {
		// 入堆
		heap.Push(&m.qMin, num)
		// 堆顶最大值
		// 不能直接pop 因为第一个元素加入一定是qMin 中，但是Pop 会出错
		// b := heap.Pop(&m.qMin).(int)
		// heap.Push(&m.qMax, b)

		// 最多相差一个值
		if m.qMax.Len()+1 < m.qMin.Len() {
			b := heap.Pop(&m.qMin).(int)
			heap.Push(&m.qMax, b)
		}

	} else {
		heap.Push(&m.qMax, num)
		// 堆顶最小值
		if m.qMax.Len() > m.qMin.Len() {
			s := heap.Pop(&m.qMax).(int)
			heap.Push(&m.qMin, s)
		}

	}

}

// FindMedian 找到中位数
// 如果总元素个数是奇数，qMin 中的数的数量比 qMax 多一个，此时中位数为 queMin 队头
// 如果是偶数，(qMin 中的最大值 + qMax 中的最小值) /2
func (m *MedianFinder) FindMedian() float64 {
	// 奇数
	if m.qMin.Len() > m.qMax.Len() {
		return float64(m.qMin[0])
	}

	return float64(m.qMax[0]+m.qMin[0]) / 2
}

// 小根堆
// []int{sum, i,j} => 两个切片的和，对应的下标
type sHeap []int

func (h sHeap) Less(i, j int) bool {
	// > 大根堆， < 小根堆
	return h[i] < h[j]
}

func (h sHeap) Len() int {
	return len(h)
}

// Swap 不会改变切片大小，所以用值接受者和指针接受者，效果相同
func (h *sHeap) Swap(i, j int) {
	(*h)[i], (*h)[j] = (*h)[j], (*h)[i]
}

func (h *sHeap) Push(x interface{}) {
	*h = append(*h, x.(int))
}

func (h *sHeap) Pop() interface{} {

	elem := (*h)[len(*h)-1]

	*h = (*h)[:len(*h)-1]
	return elem
}

File: rust/src/heap/_295_median_finder.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;

/// 数据流的中位数
/// 中位数将数据分为两个部分，左边的元素比中位数小，右边的元素比其大
/// 用两个堆q_min 来存储左边的元素， q_max 存储右边的元素
/// q_min 用大根堆，能直接拿到其最大值，q_max 用小根堆能直接拿到其最小值
/// 如果是奇数，q_min 比q_max 多一个元素(且最多只能多一个元素)，那么q_min 堆顶元素就是中位数
/// 如果是偶数，q_min 的堆顶和q_max 的堆顶是中位数
/// 需要注意的是: add_num 的时候，需要平衡左右堆的元素数量。if num < q_min[0] 较小，入左侧
/// 此刻需要判断元素数量是不是相差超过1，if yes, q_min 堆顶移到 q_max 中
/// 如果q_max 的元素数量大于q_min, 则需要平衡
struct MedianFinder {
    q_min: BinaryHeap<i32>,          // 大根堆
    q_max: BinaryHeap<Reverse<i32>>, // 小根堆
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MedianFinder {
    fn new() -> Self {
        Self {
            q_min: BinaryHeap::new(),
            q_max: BinaryHeap::new(),
        }
    }

    // 添加元素，注意平衡左右堆的元素数量
    fn add_num(&mut self, num: i32) {
        // 如果是空， 先加入q_min 中  || 小于左侧元素的最大值
        if self.q_min.is_empty() || num < *self.q_min.peek().unwrap() {
            self.q_min.push(num);

            // 元素相差大于1
            if self.q_min.len() - self.q_max.len() > 1 {
                if let Some(big) = self.q_min.pop() {
                    self.q_max.push(Reverse(big));
                }
            }
        } else {
            self.q_max.push(Reverse(num));

            if self.q_max.len() > self.q_min.len() {
                if let Some(small) = self.q_max.pop() {
                    self.q_min.push(small.0);
                }
            }
        }
    }

    fn find_median(&self) -> f64 {
        // 奇数
        if self.q_min.len() > self.q_max.len() {
            let peek = self.q_min.peek().unwrap();
            return *peek as f64;
        }
        // 偶数
        if let (Some(ele1), Some(ele2)) = (self.q_min.peek(), self.q_max.peek()) {
            return (*ele1 as f64 + ele2.0 as f64) / 2.0;
        }
        0.0
    }
}
